// project/authBehavior.js
/**
 * auth
 * Date: 23.03.14
 * Vitaliy V. Makeev (w.makeev@gmail.com)
 */
var base64 = require('project/base64');

function getBasicAuthHttpHeader(login, password) {
    return 'Basic ' + base64.encode(login + ':' + password);
}

var AuthBehavior = function () {
    var _authStore;

    this.setAuthStore = function (storeInstance) {
        _authStore = storeInstance;
    };

    this.setAuth = function (login, password) {
        if (!_authStore) { throw new Error('Auth store not defined') }
        _authStore.setAuth(login, password);
        return this;
    };

    this.getAuth = function () {
        if (!_authStore) { throw new Error('Auth store not defined') }
        return _authStore.getAuth();
    };

    this.haveAuth = function () {
        if (!_authStore) { throw new Error('Auth store not defined') }
        return _authStore.haveAuth();
    };

    this.getBasicAuthHeader = function () {
        if (!_authStore) { throw new Error('Auth store not defined') }
        var auth = this.getAuth();
        return getBasicAuthHttpHeader(auth.login, auth.password);
    };
};

module.exports = AuthProvider;
