// project/fetch/gs.js
/**
 * Google Script Http request provider factory
 * Date: 11.01.14
 * Vitaliy V. Makeev (w.makeev@gmail.com)
 */

var _ = require('lodash')
    , callbackAdapter = require('./../../../tools/callbackAdapter')
    , log = require('project/logger');

var fetch = {

    fetch: function (options, callback) {

        var _options = {
            //contentType: 'application/x-www-form-urlencoded',
            method: 'GET',
            muteHttpExceptions: true
        };
        _.extend(_options, options);

        var response, httpResponse, err;

        // Show request info
        log.info([
            'http',
            _options.method,
            _options.url
        ].join(' '));

        var startTime = new Date();

        try {
            httpResponse = UrlFetchApp.fetch(_options.url, _options);
        }
        catch (e) {
            err = e;
        }

        if (!err) {
            response = {
                headers         : httpResponse.getAllHeaders(),
                contentText     : httpResponse.getContentText(),
                responseCode    : httpResponse.getResponseCode()
            };

            // Show response info
            log.info([
                'http',
                response.responseCode,
                _options.url,
                (new Date() - startTime) + 'ms ' + response.contentText.length + 'b'
            ].join(' '));
        }

        var result = {
            response: response,
            request: _options
        };

        return callbackAdapter(err, result, callback);
    }
};

module.exports = fetch;
