// project/fetch/node.js
/**
 * Default Http request provider factory
 * Date: 11.01.14
 * Vitaliy V. Makeev (w.makeev@gmail.com)
 */

var _                 = require('lodash')
  , colors            = require('colors')
  , request           = require('request')
  , log               = require('project/logger')
  , callbackAdapter   = require('./../../../tools/callbackAdapter');

var fetch = {

    fetch: function (options, callback) {

        var _options = {
            url: options.url,
            method: options.method || 'GET',
            headers: {
                'Content-Type': options.contentType || 'application/x-www-form-urlencoded'
            }
        };

        _.extend(_options.headers, options.headers);

        if (_options.payload) _options.body = _options.payload;

        // Show request info
        log.info([
            colors.blue('http'),
            colors.magenta(_options.method),
            _options.url
        ].join(' '));

        var startTime = new Date();

        request(_options, function (error, resp, body) {
            var response;

            if (!error) {
                response = {
                    headers:        resp.headers,
                    responseCode:   resp.statusCode,
                    contentText:    body
                };

                // Show response info
                log.info([
                    colors.blue('http'),
                    colors.magenta(response.responseCode),
                    _options.url,
                    colors.grey((new Date() - startTime) + 'ms ' + response.contentText.length + 'b')
                ].join(' '));
            }

            callback(error, {
                response: response,
                request: _options
            })
        });
        
    }
};

module.exports = fetch;
