// project/fetch/node.js
/**
 * Default Http request provider factory
 * Date: 11.01.14
 * Vitaliy V. Makeev (w.makeev@gmail.com)
 */

var _                 = require('lodash'),
    httpRequest       = require('./node-https-request'),
    Queue             = require('./queue');

var queue = new Queue();

var fetch = {

    fetch: function (options, callback) {

        ['requestPeriod', 'requestsPerPeriod', 'parallelTaskCount']
            .forEach(function (opt) {
                if (options[opt] && queue[opt] !== options[opt]) {
                    queue[opt] = options[opt]
                }
            });

        var requestTask = function (cb) {
            var _options = {
                url: options.url,
                method: options.method || 'GET',
                headers: {
                    'Content-Type': options.contentType || 'application/x-www-form-urlencoded'
                }
            };

            _.extend(_options.headers, options.headers);

            if (options.payload) { _options.body = options.payload; }

            httpRequest(_options, function (err, resp, body) {
                var response;

                if (!err) {
                    response = {
                        headers: resp.headers,
                        responseCode: resp.statusCode,
                        contentText: body
                    };
                }

                cb(err, {
                    response: response,
                    request: _options
                })
            })
        };

        queue.addTask(requestTask, callback)
    }
};

module.exports = fetch;
