// project/tools/index.js
/**
 * index
 * Date: 14.06.14
 * Vitaliy V. Makeev (w.makeev@gmail.com)
 */

module.exports = {

    // Entity tools
    clone                   : require('./clone'),
    getAttr                 : require('./getAttr'),
    hasAttr                 : require('./hasAttr'),
    getPrice                : require('./getPrice'),
    hasPrice                : require('./hasPrice'),
    getPositions            : require('./getPositions'),

    // Common tools
    isUuid                  : require('./isUuid'),
    getProperty             : require('./getProperty'),


    getUriTypeName          : require('./getUriTypeName'),
    //getAttrValue            : require('./getAttrValue'),
    getPriceValue           : require('./getPriceValue'),
    //instanceOf              : require('./instanceOf'),
    //reserve                 : require('./reserve'),
    description             : require('./description'),

    format: {
        printAmount         : require('./format/moneytostr').printAmount
    }

};
