// project/tools/index.js
/**
 * index
 * Date: 14.06.14
 * Vitaliy V. Makeev (w.makeev@gmail.com)
 */

var log = require('project/logger');

module.exports = {

    getUriTypeName          : require('./getUriTypeName'),
    getAttr                 : require('./getAttr'),
    getAttrValue            : require('./getAttrValue'),
    getSalePrice            : require('./getSalePrice'),
    getSalePriceValue       : require('./getSalePriceValue'),
    setPrice                : require('./setPrice'),
    createAttributeValue    : require('./createAttrValue'), // deprecated
    createAttrValue         : require('./createAttrValue'),
    getPositions            : require('./getPositions'),
    getType                 : require('./getType'),
    getProperty             : require('./getProperty'),
    instanceOf              : require('./instanceOf'),
    clone                   : require('./clone'),
    reserve                 : require('./reserve'),
    description             : require('./description'),

    format: {
        printAmount          : require('./../moneytostr').printAmount
    },

    // [depricated] Для обратной совместимости
    getPrice            : require('./getSalePrice'),
    getPriceValue       : require('./getSalePriceValue')
};
