// project/tools/index.js
/**
 * index
 * Date: 14.06.14
 * Vitaliy V. Makeev (w.makeev@gmail.com)
 */

module.exports = {

    getUriTypeName          : require('./getUriTypeName'),
    getAttr                 : require('./getAttr'),
    getAttrValue            : require('./getAttrValue'),
    getPrice                : require('./getPrice'),
    getPriceValue           : require('./getPriceValue'),
    createAttributeValue    : require('./createAttrValue'), // deprecated
    createAttrValue         : require('./createAttrValue'),
    getPositions            : require('./getPositions'),
    getType                 : require('./getType'),
    getProperty             : require('./getProperty'),
    instanceOf              : require('./instanceOf'),
    clone                   : require('./clone'),
    reserve                 : require('./reserve'),
    description             : require('./description'),

    format: {
        printAmount          : require('./../moneytostr').printAmount
    }

    //:              require('./'),
    //:              require('./'),
    //:              require('./'),
    //:              require('./'),
    //:              require('./'),
};
