// project/model-extensions/index.js
/**
 * model-extension
 * Date: 24.05.15
 * Vitaliy V. Makeev (w.makeev@gmail.com)
 */

var _ = require('lodash');

var extensions = {
    'moysklad.__common'                 : require('./moysklad/__common'),
    'moysklad.accountEntity'            : require('./moysklad/accountEntity'),
    'moysklad.order'                    : require('./moysklad/order'),
    'moysklad.abstractGood'             : require('./moysklad/abstractGood'),
    'moysklad.operationWithPositions'   : require('./moysklad/operationWithPositions'),
    'moysklad.attributeValue'           : require('./moysklad/attributeValue')
};

module.exports.getStampExtender = function (map, client) {

    map = {
        name        : map.name,
        enums       : map.enums,
        elementInfos: _.indexBy(map.elementInfos, 'typeInfo'),
        typeInfos   : _.indexBy(map.typeInfos, function (typeInfo) {
            return map.name + '.' + typeInfo.localName
        })
    };

    return function (localName, stamp) {
        var fullTypeName = map.name + '.' + localName;

        function applyExtension (_typeName) {
            var getExtension = extensions[_typeName];
            if (getExtension) {
                var ext = getExtension(fullTypeName, map, client);
                if (ext) {
                    if (ext.state)   stamp = stamp.state(ext.state);
                    if (ext.methods) stamp = stamp.methods(ext.methods);
                    //TODO Array of encloses
                    if (ext.enclose) stamp = stamp.enclose(ext.enclose);
                }
            }
        }

        applyExtension(fullTypeName);
        applyExtension(map.name + '.__common');

        return stamp;
    }
};
